### PR TITLE
make dtype optional

### LINKED ISSUE #367: Make dtype optional for datasets
see #347

### PR DESCRIPTION
**Pull request by bendichter**
and explain that it is to be specified for concrete classes

fix #367

### CODE REVIEW

File: docs/language/source/specification_language_description.rst [hunk 1]
```diff
@@ -850,7 +850,7 @@ The specification of a datasets is described in YAML as follows:
         neurodata_type_inc: Optional neurodata_type the group should inherit from
         quantity: Optional quantity identifier for the group (default=1).
         linkable: Boolean indicating whether the group is linkable (default=True)
-        dtype: Required string describing the data type of the dataset
+        dtype: Optional string describing the data type of the dataset
```
Comment (oruebel) [inline docs/language/source/specification_language_description.rst:853]: We should add "(default=null)" to clarify that when dtype is not specified that this is the same as ``dtype: null``.

File: docs/language/source/specification_language_description.rst [hunk 2]
```diff
@@ -921,7 +921,7 @@ Boolean describing whether the this group can be linked.
 ``dtype``
 ^^^^^^^^^
 
-String describing the data type of the dataset. Same as for attributes. See :numref:`sec-dtype` for details.
+String describing the data type of the dataset. Same as for attributes. See :numref:`sec-dtype` for details. ``dtype`` may be omitted for abstract classes. Best practice is to define ``dtype`` for most concrete classes.
```
Comment (oruebel) [inline docs/language/source/specification_language_description.rst:925]: Minor clarification for the last two sentences. 

"``dtype`` may be omitted or set to ``null`` for abstract classes, indicating that the dataset or attribute may take on any valid data type. Best practice is to define ``dtype`` for concrete classes."

**Comment by bendichter** [comment]
@oruebel As far as I can tell,`dtype: null` is not currently used anywhere in the NWB schema. Why do you want to add this to the language?

**Comment by oruebel** [comment]
I don't think this is adding to the language, but as far as I can tell, this is the default behavior (at least in PyNWB/HDMF) when dtype is omitted. I guess it is Ok not mention it, since we don't use this behavior.

**Review by oruebel — APPROVED** [review]

**Review by rly — APPROVED** [review]

**Pull request merged**